File: labs/01-hello-model/python/solution/04_tool_loop.py
"""Lab 1 · step 4 — the tool-call loop, by hand. This loop is the foundation of every agent you will see today.

model -> "please call get_weather(city='Paris')" -> YOUR code runs it -> result goes back as a "tool" message ->
model answers (or asks for another tool). Frameworks (Agent Framework, LangChain, ...) run exactly this loop for you.
"""
import json
import os

from dotenv import find_dotenv, load_dotenv
from pathlib import Path as _Path
from openai import OpenAI

_here = _Path(globals().get("__file__", ".")).resolve()   # no __file__ in a notebook: start at the cwd
for _folder in [*_here.parents, _Path.cwd(), *_Path.cwd().parents]:
    if (_folder / ".env").is_file():   # every .env from this file upward: the nearest one wins,
        load_dotenv(_folder / ".env")  # so labs/.env or one .env at the repo root both work
BASE, KEY, MODEL = os.environ["GENAI_BASE_URL"], os.environ["GENAI_API_KEY"], os.environ["GENAI_MODEL"]
HEADERS = {"api-key": KEY} if os.getenv("GENAI_KEY_HEADER", "api-key") == "api-key" else None
client = OpenAI(base_url=BASE, api_key=KEY, default_headers=HEADERS)
# Optional: GENAI_REASONING_EFFORT=low (gpt-5 family) or none (local "thinking" models) keeps answers fast and cheap
EXTRA = {"reasoning_effort": os.environ["GENAI_REASONING_EFFORT"]} if os.getenv("GENAI_REASONING_EFFORT") else {}

FAKE_WEATHER = {"paris": (19, "light rain"), "mannheim": (22, "sunny"), "jacksonville": (31, "thunderstorms")}


def get_weather(city: str) -> str:
    temp, sky = FAKE_WEATHER.get(city.lower(), (20, "unknown"))
    return json.dumps({"city": city, "temp_c": temp, "sky": sky})


def to_fahrenheit(celsius: float) -> str:
    return json.dumps({"celsius": celsius, "fahrenheit": round(celsius * 9 / 5 + 32, 1)})


TOOLS = [
    {"type": "function", "function": {
        "name": "get_weather", "description": "Current weather for a city.",
        "parameters": {"type": "object", "properties": {"city": {"type": "string"}}, "required": ["city"]}}},
    {"type": "function", "function": {
        "name": "to_fahrenheit", "description": "Convert a temperature from Celsius to Fahrenheit.",
        "parameters": {"type": "object", "properties": {"celsius": {"type": "number"}}, "required": ["celsius"]}}},
]
IMPLEMENTATIONS = {"get_weather": get_weather, "to_fahrenheit": to_fahrenheit}

messages = [{"role": "system", "content": "Use the tools for facts. Be brief."},
            {"role": "user", "content": "What's the weather in Paris, and what is that temperature in Fahrenheit?"}]

for step in range(6):                                   # a budget: agents need a stop condition
    response = client.chat.completions.create(model=MODEL, messages=messages, tools=TOOLS, **EXTRA)
    msg = response.choices[0].message
    # >>> TODO 2: append the assistant message; if it has no tool_calls print the answer and stop; otherwise run each tool and append a {"role": "tool", ...} message per call
    messages.append(msg.model_dump(exclude_none=True))  # the assistant turn (incl. its tool_calls) stays in the history
    if not msg.tool_calls:
        print(f"\nANSWER after {step} tool round(s): {msg.content}")
        break
    for call in msg.tool_calls:
        args = json.loads(call.function.arguments or "{}")
        result = IMPLEMENTATIONS[call.function.name](**args)
        print(f"  tool call #{step + 1}: {call.function.name}({args}) -> {result}")
        messages.append({"role": "tool", "tool_call_id": call.id, "content": result})
    # <<< TODO
else:
    print("Stopped: step budget exhausted.")

# Try this:
#  * Ask something that needs no tool ("Tell me a joke") — the model answers directly.
#  * Ask for three cities at once — many models return several tool_calls in ONE turn (parallel tool calls).
#  * Remove the system prompt or a tool description and watch the tool choice get worse.

File: labs/01-hello-model/csharp/solution/04_tool_loop.cs
#:package OpenAI@2.13.0
#:package DotNetEnv@3.2.0
#:property PublishAot=false
// Lab 1 · step 4 — the tool-call loop, by hand. This loop is the foundation of every agent you will see today.
//
// model -> "please call get_weather(city='Paris')" -> YOUR code runs it -> result goes back as a "tool" message ->
// model answers (or asks for another tool). Frameworks run exactly this loop for you: M.E.AI's FunctionInvokingChatClient
// and every Agent Framework agent (Lab 3) — here we write it ourselves with the plain OpenAI .NET SDK.
using System.ClientModel;
using System.ClientModel.Primitives;
using System.Text.Json;
using OpenAI;
using OpenAI.Chat;
#pragma warning disable OPENAI001 // ReasoningEffortLevel is still marked "experimental" in the OpenAI .NET SDK 2.13

var labs = AppContext.GetData("EntryPointFileDirectoryPath") as string ?? Directory.GetCurrentDirectory();
while (!File.Exists(Path.Combine(labs, "_tools", "make_starters.py")))    // walk up to the labs/ root
    labs = Path.GetDirectoryName(labs) ?? throw new DirectoryNotFoundException("labs/_tools not found above this file");
// labs/.env, or one .env at the repo root: load every .env from labs/ upward - NoClobber, so the nearest wins
for (var envDir = new DirectoryInfo(labs); envDir is not null; envDir = envDir.Parent)
    if (File.Exists(Path.Combine(envDir.FullName, ".env"))) DotNetEnv.Env.NoClobber().Load(Path.Combine(envDir.FullName, ".env"));   // labs/.env; real environment variables win
static string Env(string name, string? fallback = null) => Environment.GetEnvironmentVariable(name) is { Length: > 0 } value
    ? value : fallback ?? throw new InvalidOperationException($"{name} is not set - copy labs/.env.example to labs/.env");

string key = Env("GENAI_API_KEY"), model = Env("GENAI_MODEL");
var clientOptions = new OpenAIClientOptions { Endpoint = new Uri(Env("GENAI_BASE_URL")) };
if ((Environment.GetEnvironmentVariable("GENAI_KEY_HEADER") ?? "api-key") == "api-key")
    clientOptions.AddPolicy(new HeaderPolicy("api-key", key), PipelinePosition.PerCall);
ChatClient client = new OpenAIClient(new ApiKeyCredential(key), clientOptions).GetChatClient(model);

var fakeWeather = new Dictionary<string, (int TempC, string Sky)>
{
    ["paris"] = (19, "light rain"), ["mannheim"] = (22, "sunny"), ["jacksonville"] = (31, "thunderstorms"),
};

string GetWeather(string city)
{
    var (temp, sky) = fakeWeather.GetValueOrDefault(city.ToLowerInvariant(), (20, "unknown"));
    return JsonSerializer.Serialize(new { city, temp_c = temp, sky });
}

string ToFahrenheit(double celsius) => JsonSerializer.Serialize(new { celsius, fahrenheit = Math.Round(celsius * 9 / 5 + 32, 1) });

// The tool descriptions the model sees: name, description, JSON Schema of the arguments.
var options = new ChatCompletionOptions
{
    Tools =
    {
        ChatTool.CreateFunctionTool("get_weather", "Current weather for a city.", BinaryData.FromString(
            """{"type": "object", "properties": {"city": {"type": "string"}}, "required": ["city"]}""")),
        ChatTool.CreateFunctionTool("to_fahrenheit", "Convert a temperature from Celsius to Fahrenheit.", BinaryData.FromString(
            """{"type": "object", "properties": {"celsius": {"type": "number"}}, "required": ["celsius"]}""")),
    },
};
// Optional: GENAI_REASONING_EFFORT=low (gpt-5 family) or none (local "thinking" models) keeps answers fast and cheap
if (Env("GENAI_REASONING_EFFORT", "") is { Length: > 0 } effort) options.ReasoningEffortLevel = new ChatReasoningEffortLevel(effort);

// Our side of the contract: run the function the model asked for, with the arguments it sent (a JSON string).
string Run(ChatToolCall call)
{
    using JsonDocument args = JsonDocument.Parse(call.FunctionArguments);
    return call.FunctionName switch
    {
        "get_weather" => GetWeather(args.RootElement.GetProperty("city").GetString()!),
        "to_fahrenheit" => ToFahrenheit(args.RootElement.GetProperty("celsius").GetDouble()),
        _ => $$"""{"error": "unknown tool {{call.FunctionName}}"}""",
    };
}

List<ChatMessage> messages =
[
    new SystemChatMessage("Use the tools for facts. Be brief."),
    new UserChatMessage("What's the weather in Paris, and what is that temperature in Fahrenheit?"),
];

const int Budget = 6;                                   // a budget: agents need a stop condition
int step = 0;
for (; step < Budget; step++)
{
    ChatCompletion response = await client.CompleteChatAsync(messages, options);
    // >>> TODO 2: append the assistant message; if it has no ToolCalls print the answer and stop; otherwise run each tool and append a ToolChatMessage per call
    messages.Add(new AssistantChatMessage(response));    // the assistant turn (incl. its tool calls) stays in the history
    if (response.ToolCalls.Count == 0)
    {
        Console.WriteLine($"\nANSWER after {step} tool round(s): {response.Content[0].Text}");
        break;
    }
    foreach (ChatToolCall call in response.ToolCalls)
    {
        string result = Run(call);
        Console.WriteLine($"  tool call #{step + 1}: {call.FunctionName}({call.FunctionArguments}) -> {result}");
        messages.Add(new ToolChatMessage(call.Id, result));
    }
    // <<< TODO
}
if (step == Budget) Console.WriteLine("Stopped: step budget exhausted.");

// Try this:
//  * Ask something that needs no tool ("Tell me a joke") — the model answers directly.
//  * Ask for three cities at once — many models return several ToolCalls in ONE turn (parallel tool calls).
//  * Remove the system prompt or a tool description and watch the tool choice get worse.

sealed class HeaderPolicy(string name, string value) : PipelinePolicy   // adds the key header to every request
{
    public override void Process(PipelineMessage message, IReadOnlyList<PipelinePolicy> pipeline, int index)
    { message.Request.Headers.Set(name, value); ProcessNext(message, pipeline, index); }
    public override ValueTask ProcessAsync(PipelineMessage message, IReadOnlyList<PipelinePolicy> pipeline, int index)
    { message.Request.Headers.Set(name, value); return ProcessNextAsync(message, pipeline, index); }
}
